File: javascript/templates/dfs_binary_tree.js
class TreeNode {
    constructor(val) {
        this.value = val;
        this.left = null;
        this.right = null;
    }
}

// Function to perform Depth-First Search (DFS) - Inorder Traversal
function dfsInorder(node) {
    if (node !== null) {
        dfsInorder(node.left);
        console.log(node.value + " ");
        dfsInorder(node.right);
    }
}

// Function to perform Depth-First Search (DFS) - Preorder Traversal
function dfsPreorder(node) {
    if (node !== null) {
        console.log(node.value + " ");
        dfsPreorder(node.left);
        dfsPreorder(node.right);
    }
}

// Function to perform Depth-First Search (DFS) - Postorder Traversal
function dfsPostorder(node) {
    if (node !== null) {
        dfsPostorder(node.left);
        dfsPostorder(node.right);
        console.log(node.value + " ");
    }
}

// Example usage:
// Build a sample binary tree:
//         1
//        / \
//       2   3
//      / \
//     4   5

const root = new TreeNode(1);
root.left = new TreeNode(2);
root.right = new TreeNode(3);
root.left.left = new TreeNode(4);
root.left.right = new TreeNode(5);

console.log("Inorder Traversal (DFS): ");
dfsInorder(root);

console.log("Preorder Traversal (DFS): ");
dfsPreorder(root);

console.log("Postorder Traversal (DFS): ");
dfsPostorder(root);

File: python/templates/dfs_binary_tree.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

# Function to perform Depth-First Search (DFS) - Inorder Traversal
def dfs_inorder(node):
    if node is not None:
        dfs_inorder(node.left)
        print(node.value, end=' ')
        dfs_inorder(node.right)
    else:
        return

# Function to perform Depth-First Search (DFS) - Preorder Traversal
def dfs_preorder(node):
    if node is not None:
        print(node.value, end=' ')
        dfs_preorder(node.left)
        dfs_preorder(node.right)
    else:
        return

# Function to perform Depth-First Search (DFS) - Postorder Traversal
def dfs_postorder(node):
    if node is not None:
        dfs_postorder(node.left)
        dfs_postorder(node.right)
        print(node.value, end=' ')
    else:
        return

# Example usage:
if __name__ == "__main__":
    # Build a sample binary tree:
    #         1
    #        / \
    #       2   3
    #      / \
    #     4   5

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)

    print("Inorder Traversal (DFS):")
    dfs_inorder(root)
    print()

    print("Preorder Traversal (DFS):")
    dfs_preorder(root)
    print()

    print("Postorder Traversal (DFS):")
    dfs_postorder(root)
    print()